platform/o.n.bootstrap/src/org/netbeans/JarClassLoader.java

```diff
@@ -574,13 +609,81 @@ protected URL doGetResource(String name) throws IOException  {
         @Override
         protected byte[] readClass(String path) throws IOException {
             try {
+                if (isMultiRelease() && RUNTIME_VERSION > BASE_VERSION) {
+                    int[] vers = getVersions();
+                    if (vers.length > 0) {
+                        for (int i = vers.length - 1; i >= 0; i--) {
+                            int version = vers[i];
+                            if (version > RUNTIME_VERSION) {
+                                continue;
+                            }
+                            if (version < BASE_VERSION) {
+                                break;
+                            }
+                            byte[] data = archive.getData(this, "META-INF/versions/" + version + "/" + path);
+                            if (data != null) {
+                                return data;
+                            }
+                        }
+                    }
+                }
                 return archive.getData(this, path);
             } catch (ZipException ex) {
                 dumpFiles(file, -1);
                 throw ex;
             }
         }
-        
+
+        private int[] getVersions() {
+            int[] ret;
+            if (versions != null && (ret = versions.get()) != null) {
+                return ret;
+            }
+            try {
+                JarFile src = getJarFile("versions");
```

Comment:
is "versions" correct here? Shouldn't this be a path within the jar?

Reply:
Agreed looks like "versions" is just the name of the jar in the test case.